[lambraydon] iP

lambraydon opened the pull request:
# Ben
> "If you like to use CLI's, try out Ben" - Einstein, Probably ([Source](https://www.youtube.com/results?search_query=rick+roll))

Ben frees your mind of having to remember things you need to do. It's,

* text-based
* easy to learn
* ~~FAST~~ SUPER FAST to use

All you need to do is,
1. add your tasks.
2. let it manage your tasks for you 😉
3. And it is **FREE**!

Features:

 - [X] Managing tasks
 - [ ] Managing deadlines (coming soon)
 - [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the main method:

```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

lambraydon committed:
Modified chatbot name to Ben. Added Ben.java and deleted Duke.java. Skeletal version greets the user and exits immediately

lambraydon committed:
Added echo and bye command

lambraydon committed:
Added Task.java. Added list functionality for Ben.java

lambraydon committed:
Added isCompleted fields to Task.java. Added mark and unmark conditions in Ben.java

lambraydon committed:
Added Deadlines.java, Events.java and Todos.java. Chatbot can only take in these 3 add commands

lambraydon committed:
Reformated output of addSuccessMessage() method in Ben.java to include horizontal lines

lambraydon committed:
Added automated tests in input.txt and expected results in EXPECTED.TXT

lambraydon committed:
Added InvalidCommandException.java and EmptyDescriptionException.java to handle invalid command and empty description errors. Updated test scripts to account for these errors

lambraydon committed:
Added delete as one of the editList commands in Ben.java. Updated test files to include delete test case

lambraydon committed:
Add TaskList.java. TaskList.java encapsulates all task operations previously in Ben.java

lambraydon committed:
Add data/ben.txt. Added saveTasks and loadTasks method in Ben.java to save and load task list to ben.txt

lambraydon committed:
Merge branch 'branch-Level-7'

lambraydon committed:
Modify Tasks to accept dateTime objects instead of string representation of dates

lambraydon committed:
Merge branch 'branch-Level-8'

lambraydon committed:
Add Storage.java. Storage.java encapsulates methods that deal with loading and saving tasks from file

lambraydon committed:
Add Parser.java. Parser.java encapsulates methods that deal with making sense of user command

lambraydon committed:
Modify Parser.java to include isEditListCommand method

lambraydon committed:
Add Ui.java. Ui.java encampsulates methods that deal with interactions with the user

lambraydon committed:
Merge branch 'A-MoreOOP'

lambraydon committed:
Add package ben

lambraydon committed:
Merge branch 'A-Packages'

lambraydon committed:
Modify Ui.java to include scanner

lambraydon committed:
Fix merge conflicts with master

lambraydon committed:
Add Command files. Extracted commands as classes

lambraydon committed:
Merge branch 'A-MoreOOP'

lambraydon committed:
Merge remote-tracking branch 'origin/add-gradle-support' into A-Gradle

lambraydon committed:
Merge branch 'A-Gradle'

lambraydon committed:
Add AddCommandTest.java and ParserTest.java

lambraydon committed:
Merge branch 'A-JUnit'

lambraydon committed:
Add JavaDocs to all classes

lambraydon committed:
Add FindCommand.java

lambraydon committed:
Modify package ben to adhere to coding style

lambraydon committed:
Modify FindCommand.java to include javadocs

lambraydon committed:
Merge branch 'A-JavaDoc'

lambraydon committed:
Resolve merge conflicts

lambraydon committed:
Resolve merge conflicts

nixonwidjaja commented on src/test/java/ben/ParserTest.java:
```diff
@@ -0,0 +1,27 @@
+
+public class ParserTest {
+    @Test
+    public void invalidCommandTest(){
```
Perhaps you can follow featureUnderTest_testScenario_expectedBehavior() for test name?

Sheeepen replied:
I concur! A description for explaining how it is invalid would be very helpful

nixonwidjaja reviewed (changes requested):
Overall great job on creating the abstraction and encapsulation of each class. I like how you use screaming snake case for static variables which many others missed. Some issue for naming convention for tests can be easily fixed. Perhaps consider packaging different responsibilities into different folders? e.g. task, command, parser, etc folders.

lambraydon committed:
Add GUI

lambraydon committed:
Merge branch 'branch-Level-10'

andrefoo commented:
> It seems like your classes are not packaged, and all of them fall under the same 'ben' package. It would be helpful to a reader and someone that might take on your project in the future to be able to see the code in neatly separated packages. For example, you could separate the classes like this:

'''
ben/
│
├── command/ # All command-related classes
│ ├── Command.java
│ ├── AddCommand.java
│ ├── DeleteCommand.java
│ ├── MarkCommand.java
│ ├── UnmarkCommand.java
│ └── ... (other command classes)
│
├── storage/ # All storage-related classes
│ └── Storage.java
│
├── task/ # Task and related subclasses
│ ├── Task.java
│ ├── ToDo.java
│ ├── Deadline.java
│ ├── Event.java
│ └── ... (other task-related classes)
│
├── ui/ # User interface classes
│ └── Ui.java
│
├── parser/ # Parsing-related classes
│ └── Parser.java
│
└── Ben.java # Main Duke application class
'''

Sheeepen commented on src/main/java/ben/Ben.java:
```diff
@@ -0,0 +1,66 @@
+/**
+ * Represents a chatbot named Ben.
+ */
+public class Ben {
```
Cute name! Reminds me of my cousin named Ben Dover 😄

Sheeepen commented on src/main/java/ben/ByeCommand.java:
```diff
@@ -0,0 +1,27 @@
+package ben;
+
+/**
+ * Represents a bye command.
```
Very minor nit but perhaps capitalise your classes? (i.e. Bye)

Sheeepen commented on src/main/java/ben/Events.java:
```diff
@@ -0,0 +1,52 @@
+/**
+ * Represents an events task.
+ */
+public class Events extends Task{
```
Consulting the Java coding standards, perhaps using singular Event (without the s) is more accurate to represent a single Event instance. Events make it sounds like a collection class.

Sheeepen replied:
Same for the ToDos and Deadlines!

Sheeepen reviewed (commented):
LGTM. Mostly nits. 
One of the easiest improvements I think you could add for code readability would be to package your classes into commands, exceptions, tasks etc. instead of putting them all under ben.

lambraydon committed:
Add assertions

Methods in ben make assumptions about the arguments parsed.

If arguments parsed do not fit this assumption, the method does not account for this resulting in a runtime error.

Assertions are used to make sure the assumptions are true.

lambraydon committed:
Modify ben: increase code quality

Goes against code quality principles.

Rename classes to make it singular so that it does no sound like a collection. Modify if-else blocks.

Keep code readable.

lambraydon committed:
Add Feature: Friendlier Syntax

Command syntax is only allows for commands to be written fully and is non-case sensitive.

Let's make the command syntax more user-friendly and flexible by providing command syntax shortcuts e.g "td" is a shorter alias for "todo".

Creating shortcuts allows user to type the commands faster and improves efficiency.

lambraydon committed:
Merge pull request #4 from lambraydon/branch-A-Assertions

Add assertions

lambraydon committed:
Merge branch 'master' into branch-A-CodeQuality

lambraydon committed:
Merge branch 'master' into branch-BCD-Extension

lambraydon committed:
Merge pull request #3 from lambraydon/branch-BCD-Extension

Add Feature: Friendlier Syntax

lambraydon committed:
Merge branch 'master' into branch-A-CodeQuality

lambraydon committed:
Merge pull request #2 from lambraydon/branch-A-CodeQuality

Modify ben: increase code quality

lambraydon committed:
Delete: Hello.java

lambraydon committed:
Modify GUI: Add padding and color

GUI does not have color and padding.

UI looks unplesant, user is unable to distinguish between message sent and message from ben as the output.

We can add different background colors and padding to diffenrentiaite between dialog boxes.

lambraydon committed:
Add User Guide

lambraydon committed:
Modify header comments in Ben

lambraydon committed:
Modify User Guide

lambraydon committed:
Merge branch 'branch-A-UserGuide'

lambraydon committed:
Modify Storage.java to creat file if it does not exist

lambraydon committed:
Modify ben to include sub-packages

lambraydon committed:
Modify filePath for data
